src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,908 @@
+            
+            filter(
+                static_cast<float>(tile_pixel_x + sample_pos_x),
+                static_cast<float>(tile_pixel_y + sample_pos_y),
+                tile_crop_window,
+                m3hash);
+        }
+
+    private:
+        const Filter2f*                 m_renderer_filter;
+        int                             m_tile_origin_x;
+        int                             m_tile_origin_y;
+        int                             m_tile_end_x;
+        int                             m_tile_end_y;
+        size_t                          m_frame_width;
+        double                          m_rcp_canvas_width;
+        double                          m_rcp_canvas_height;
+        AABB2u                          m_crop_window;
+        Image&                          m_aov_image;
+        size_t                          m_num_layers;
+        vector<WeightMap>&              m_pixel_samples;
+        map<pair<int, int>, NameMap>&   m_tile_name_map;
+        CryptomatteAOV::CryptomatteType m_layer_type;
+
+        // Copy of the FilteredTile::add().
+        void filter(
+            const float                 x,
+            const float                 y,
+            const AABB2u&               crop_window,
+            const uint32                m3hash)
+        {
+            assert(m_renderer_filter);
+
+            // Convert (x, y) from continuous image space to discrete image space.
+            const float dx = x - 0.5f;
+            const float dy = y - 0.5f;
+
+            // Find the pixels affected by this sample.
+            AABB2i footprint;
+            footprint.min.x = truncate<int>(fast_ceil(dx - m_renderer_filter->get_xradius()));
+            footprint.min.y = truncate<int>(fast_ceil(dy - m_renderer_filter->get_yradius()));
+            footprint.max.x = truncate<int>(fast_floor(dx + m_renderer_filter->get_xradius()));
+            footprint.max.y = truncate<int>(fast_floor(dy + m_renderer_filter->get_yradius()));
+
+            // Don't affect pixels outside the crop window.
+            footprint = AABB2i::intersect(footprint, crop_window);
+
+            // Bail out if the point does not fall inside the crop window.
+            // Only check the x coordinate; the y coordinate is checked in the loop below.
+            if (footprint.min.x > footprint.max.x)
+                return;
+
+            for (int ry = footprint.min.y; ry <= footprint.max.y; ++ry)
+            {
+                for (int rx = footprint.min.x; rx <= footprint.max.x; ++rx)
+                {
+                    const float weight = m_renderer_filter->evaluate(rx - dx, ry - dy);
+                    WeightMap& weight_map = m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)];
+                    const float old_value = weight_map.get(m3hash);
+                    weight_map.insert(m3hash, old_value + weight); // possible race conditions writing to pixels on the tile edges
```

Comment:
That's a big problem!

`FilteredTile` handles it by using `foundation::atomic_add()` to atomically add weights; for that to work here, `WeightMap` needs to expose a mutable refs to values.

You can look at how `FilteredTile::atomic_add()` is implemented.